packages/next/build/swc/index.js

```diff
@@ -1,46 +1,81 @@
+import path from 'path'
+import { pathToFileURL } from 'url'
 import { platform, arch } from 'os'
 import { platformArchTriples } from 'next/dist/compiled/@napi-rs/triples'
-import { version as nextVersion, optionalDependencies } from 'next/package.json'
 import * as Log from '../output/log'
 import { getParserOptions } from './options'
 import { eventSwcLoadFailure } from '../../telemetry/events/swc-load-failure'
 import { patchIncorrectLockfile } from '../../lib/patch-incorrect-lockfile'
+import { downloadWasmSwc } from '../../lib/download-wasm-swc'
+import { version as nextVersion } from 'next/package.json'
 
 const ArchName = arch()
 const PlatformName = platform()
 const triples = platformArchTriples[PlatformName][ArchName] || []
 
 let nativeBindings
 let wasmBindings
+let downloadWasmPromise
+let pendingBindings
 export const lockfilePatchPromise = {}
 
 async function loadBindings() {
-  if (!lockfilePatchPromise.cur) {
-    // always run lockfile check once so that it gets patched
-    // even if it doesn't fail to load locally
-    lockfilePatchPromise.cur = patchIncorrectLockfile(process.cwd()).catch(
-      console.error
-    )
+  if (pendingBindings) {
+    return pendingBindings
   }
+  pendingBindings = new Promise(async (resolve, reject) => {
+    if (!lockfilePatchPromise.cur) {
+      // always run lockfile check once so that it gets patched
+      // even if it doesn't fail to load locally
+      lockfilePatchPromise.cur = patchIncorrectLockfile(process.cwd()).catch(
+        console.error
+      )
+    }
 
-  let attempts = []
-  try {
-    return loadNative()
-  } catch (a) {
-    attempts = attempts.concat(a)
-  }
+    let attempts = []
+    try {
+      return resolve(loadNative())
+    } catch (a) {
+      attempts = attempts.concat(a)
+    }
 
-  // TODO: fetch wasm and fallback when loading native fails
-  // so that users aren't blocked on this, we still want to
-  // report the native load failure so we can patch though
-  try {
-    let bindings = await loadWasm()
-    return bindings
-  } catch (a) {
-    attempts = attempts.concat(a)
-  }
+    try {
+      let bindings = await loadWasm()
+      eventSwcLoadFailure({ wasm: 'enabled' })
+      return resolve(bindings)
+    } catch (a) {
+      attempts = attempts.concat(a)
+    }
 
-  logLoadFailure(attempts)
+    try {
+      // if not installed already download wasm package on-demand
+      // we download to a custom directory instead of to node_modules
+      // as node_module import attempts are cached and can't be re-attempted
+      // x-ref: https://github.com/nodejs/modules/issues/307
+      const wasmDirectory = path.join(
+        path.dirname(require.resolve('next/package.json')),
+        'wasm'
+      )
+      if (!downloadWasmPromise) {
+        downloadWasmPromise = downloadWasmSwc(nextVersion, wasmDirectory)
```

Comment:
This gonna download when first time swc is called? maybe we could have a export api like `initSWCWasm`, and do this when build or dev start where the timing is more earlier and not very tricky?

Reply:
Yeah the first time `loadBindings` is called, was wondering if it might be good to make loading the bindings more eager although we might still want to avoid loading them when babel is being used